### PR TITLE
Update to JRuby `9.4`

### PR DESCRIPTION
**Pull request by andsel**
## What does this PR do?

- Redefine space token in `LSCL`  and `grammar` treetop from `_` which would generated methods in the form `def _0` (deprecated since `2.7`) to `sc`.
- `I18n.t` method doesn't accept hash as second argument
- ~~`URI.encode` -> `URI.encode_www_form_component`~~
  - `URI.encode` has been replaced with same functionality with `URI::Parser.new.escape`
- `YAML.load` needs explicit `fallback: false` to return false when the yaml string is empty (or contains only comments)
- JRuby's  `JavaClass` has been removed, now it can use `java.lang.Class` directly
- explicitly require gem `thwait` to satisfy `require "thwait"` (In `Gemfile.template` and `logstash-core/logstash-core.gemspec`)
- fix not args `clone` to be `def clone(*args)`
- fix `Enumeration.each_slice` which from `Ruby 3.1` is [chainable](https://rubyreferences.github.io/rubychanges/3.1.html#enumerableeach_cons-and-each_slice-return-a-receiver) and doesn't return `nil`. JRuby fixed in https://github.com/jruby/jruby/issues/7015
- Expanded `Down.download` arguments map ca16bbed3c302006967413eb9d3862f2da81f7ae
- Avoid to pass `nil` in the list of couples used in `Hash[ <list of couples> ]` which from Ruby `3.0` generates an `ArgumentError`
- Removed space not allowed between method name and parentheses `initialize (` is forbidden. 29b607dcdef98f81a73ad171639fd13aaa65e243
- With [Ruby 2.7 the `Kernel#open`](https://rubyreferences.github.io/rubychanges/2.7.html#network-and-web) doesn't fallback to `URI#open`, fixed test code that used that to verify open port. e5b70de54c5301f51a767da67294092af0cfafdc
- Avoid to drop `rdoc/` folder from vendored JRuby else `bin/logstash -i irb` would crash, commit b71f73e9c6edb81a7b7ae1305047e506f61c6e8c
```bash
LoadError: no such file to load -- rdoc/version
```

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files (and/or docker env variables)
- [ ] I have added tests that prove my fix is effective or that my feature works

## Author's Checklist

- [x] check previous update #14114
- [x] Ruby change list in [3.1](https://rubyreferences.github.io/rubychanges/3.1.html)
- [ ] check about the problem with `lib/jni` and glibc, PR to fix in LS is #14890. Verify if port to this or remove the original effort if it's resolved in `9.4.x` in the meantime.
- [x] verify `irb` works: `bin/logstah -i irb`

## Related issues

- Relates #14802

### CODE REVIEW

**Comment by andsel** [comment]
Jenkins test this please

File: rubyUtils.gradle [hunk 1]
```diff
@@ -25,7 +25,7 @@ buildscript {
     dependencies {
         classpath "org.yaml:snakeyaml:${snakeYamlVersion}"
         classpath "de.undercouch:gradle-download-task:4.0.4"
-        classpath "org.jruby:jruby-core:9.3.10.0"
+        classpath "org.jruby:jruby-complete:9.4.0.0"
```
Comment (jsvd) [inline rubyUtils.gradle:28]: do we need to reintroduce the complete jar here?

File: rubyUtils.gradle [hunk 2]
```diff
@@ -168,8 +168,8 @@ void setupJruby(File projectDir, File buildDir) {
 Object executeJruby(File projectDir, File buildDir, Closure<?> /* Object*/ block) {
     def jruby = new ScriptingContainer()
     def env = jruby.environment
-    def gemDir = "${projectDir}/vendor/bundle/jruby/2.6.0".toString()
-    jruby.setLoadPaths(["${projectDir}/vendor/bundle/jruby/2.6.0/gems/bundler-2.4.13/lib".toString(), "${projectDir}/vendor/jruby/lib/ruby/stdlib".toString()])
+    def gemDir = "${projectDir}/vendor/bundle/jruby/3.1.0".toString()
+    jruby.setLoadPaths(["${projectDir}/vendor/bundle/jruby/3.1.0/gems/bundler-2.4.13/lib".toString(), "${projectDir}/vendor/jruby/lib/ruby/stdlib".toString()])
```
Comment (jsvd) [inline rubyUtils.gradle:172]: either we apply the same change as in https://github.com/elastic/logstash/pull/15066, or need to update the path to be bundler-2.4.14, released 3 days ago https://rubygems.org/gems/bundler/versions/2.4.14

File: spec/unit/plugin_manager/pack_installer/local_spec.rb [hunk 3]
```diff
@@ -67,8 +67,8 @@
       it "install the gems" do
         expect(::Bundler::LogstashInjector).to receive(:inject!).with(be_kind_of(LogStash::PluginManager::PackInstaller::Pack)).and_return([])
 
-        expect(::LogStash::PluginManager::GemInstaller).to receive(:install).with(/logstash-input-packtest/, anything)
-        expect(::LogStash::PluginManager::GemInstaller).to receive(:install).with(/logstash-input-packtestdep/, anything)
+        expect(::LogStash::PluginManager::GemInstaller).to receive(:install).with(/logstash-input-packtest-/, anything)
```
Comment (jsvd) [inline spec/unit/plugin_manager/pack_installer/local_spec.rb:70]: why is this change necessary?

File: rubyUtils.gradle (hunk 1 above)
Comment (andsel) [reply rubyUtils.gradle:28]: The version change was done before #15050 was merged, in that PR the `-complete` to `-core` was made: https://github.com/elastic/logstash/pull/15050/files#diff-7a2156b270dee259f243cf6371cdeb28e3bfdbc8c0e418cb37cfb4724144b305L28

Comment (andsel) [reply rubyUtils.gradle]: Fixed with 8dd850ff8ac87592128239ba2c1de045a7d47f02

File: spec/unit/plugin_manager/pack_installer/local_spec.rb (hunk 3 above)
Comment (andsel) [reply spec/unit/plugin_manager/pack_installer/local_spec.rb:70]: The commit that introduces the changes is 0c92eb34055188e7dd543758b58be5128ddba2a3.

Without this change, the test fail:
```sh
SPEC_OPTS="-fd -P spec/unit/plugin_manager/pack_installer/local_spec.rb" ./gradlew clean :logstash-core:rubyTests --tests org.logstash.RSpecTests
```

with 

```
    Failures:

      1) LogStash::PluginManager::PackInstaller::Local when the local file exist when the pack is valid install the gems
         Failure/Error: expect(::LogStash::PluginManager::GemInstaller).to receive(:install).with(/logstash-input-packtest/, anything)

           (LogStash::PluginManager::GemInstaller (class)).install(/logstash-input-packtest/, anything)
               expected: 1 time with arguments: (/logstash-input-packtest/, anything)
               received: 2 times with arguments: (/logstash-input-packtest/, anything)
         # ./spec/unit/plugin_manager/pack_installer/local_spec.rb:70:in `block in <main>'
         # ./lib/bootstrap/rspec.rb:36:in `<main>'

    Finished in 0.04501 seconds (files took 0.02848 seconds to load)
    5 examples, 1 failure

    Failed examples:

    rspec ./spec/unit/plugin_manager/pack_installer/local_spec.rb:67 # LogStash::PluginManager::PackInstaller::Local when the local file exist when the pack is valid install the gems


org.logstash.RSpecTests > rspecTests[core tests] FAILED
    java.lang.AssertionError: RSpec test suite `core tests` saw at least one failure.
        at org.junit.Assert.fail(Assert.java:88)
        at org.logstash.RSpecTests.rspecTests(RSpecTests.java:72)
```

Comment (jsvd) [reply spec/unit/plugin_manager/pack_installer/local_spec.rb:70]: really strange that it worked before and now doesn't. main and this PR are using the same version of rspec, but bundler and jruby are different.

**Comment by jsvd** [comment]
> check about the problem with lib/jni and glibc, PR to fix in LS is https://github.com/elastic/logstash/pull/14890. Verify if port to this or remove the original effort if it's resolved in 9.4.x in the meantime.

We can remove the workaround:

```
[root@a03afd0551ce ~]# cat /etc/redhat-release
Red Hat Enterprise Linux Server release 7.9 (Maipo)
[root@a03afd0551ce ~]# uname -a
Linux a03afd0551ce 5.15.49-linuxkit #1 SMP PREEMPT Tue Sep 13 07:51:32 UTC 2022 x86_64 x86_64 x86_64 GNU/Linux[root@a03afd0551ce ~]# ldd jruby-9.4.2.0/lib/jni/x86_64-Linux/libjffi-1.2.so
ldd: warning: you do not have execution permission for `jruby-9.4.2.0/lib/jni/x86_64-Linux/libjffi-1.2.so'
jruby-9.4.2.0/lib/jni/x86_64-Linux/libjffi-1.2.so: /lib64/libc.so.6: version `GLIBC_2.27' not found (required by jruby-9.4.2.0/lib/jni/x86_64-Linux/libjffi-1.2.so)
	libc.so.6 => /lib64/libc.so.6 (0x00007fffff9eb000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fffffddc000)
[root@a03afd0551ce ~]# ldd jruby-9.4.3.0/lib/jni/x86_64-Linux/libjffi-1.2.so
ldd: warning: you do not have execution permission for `jruby-9.4.3.0/lib/jni/x86_64-Linux/libjffi-1.2.so'
	libc.so.6 => /lib64/libc.so.6 (0x00007fffff7ec000)
	/lib64/ld-linux-x86-64.so.2 (0x00007fffffddc000)
```

**Comment by jsvd** [comment]
Then again, maybe not, still broken on aarch64:

```
❯ docker run -it oraclelinux:7 bash
[root@5db95de36c6b /]# uname -a
Linux 5db95de36c6b 5.15.49-linuxkit #1 SMP PREEMPT Tue Sep 13 07:51:32 UTC 2022 aarch64 aarch64 aarch64 GNU/Linux
[root@5db95de36c6b /]# curl -s https://repo1.maven.org/maven2/org/jruby/jruby-dist/9.2.21.0/jruby-dist-9.2.21.0-bin.tar.gz | tar -zxf -
[root@5db95de36c6b /]# curl -s https://repo1.maven.org/maven2/org/jruby/jruby-dist/9.3.9.0/jruby-dist-9.3.9.0-bin.tar.gz | tar -zxf -
[root@5db95de36c6b /]# curl -s https://repo1.maven.org/maven2/org/jruby/jruby-dist/9.4.3.0/jruby-dist-9.4.3.0-bin.tar.gz | tar -zxf -
[root@5db95de36c6b /]# ldd jruby-9.2.21.0/lib/jni/aarch64-Linux/libjffi-1.2.so
ldd: warning: you do not have execution permission for `jruby-9.2.21.0/lib/jni/aarch64-Linux/libjffi-1.2.so'
	linux-vdso.so.1 =>  (0x0000ffff89de4000)
	libc.so.6 => /lib64/libc.so.6 (0x0000ffff89c1e000)
	/lib/ld-linux-aarch64.so.1 (0x0000ffff89db6000)
[root@5db95de36c6b /]# ldd jruby-9.3.9.0/lib/jni/aarch64-Linux/libjffi-1.2.so
ldd: warning: you do not have execution permission for `jruby-9.3.9.0/lib/jni/aarch64-Linux/libjffi-1.2.so'
	linux-vdso.so.1 =>  (0x0000ffff9bd3f000)
	libc.so.6 => /lib64/libc.so.6 (0x0000ffff9bb79000)
	/lib/ld-linux-aarch64.so.1 (0x0000ffff9bd11000)
[root@5db95de36c6b /]# ldd jruby-9.4.3.0/lib/jni/aarch64-Linux/libjffi-1.2.so
ldd: warning: you do not have execution permission for `jruby-9.4.3.0/lib/jni/aarch64-Linux/libjffi-1.2.so'
jruby-9.4.3.0/lib/jni/aarch64-Linux/libjffi-1.2.so: /lib64/libc.so.6: version `GLIBC_2.27' not found (required by jruby-9.4.3.0/lib/jni/aarch64-Linux/libjffi-1.2.so)
	linux-vdso.so.1 =>  (0x0000ffffae948000)
	libc.so.6 => /lib64/libc.so.6 (0x0000ffffae763000)
	/lib/ld-linux-aarch64.so.1 (0x0000ffffae91a000)
```

Which lines up correctly to the comment here: https://github.com/jruby/jruby/issues/7579#issuecomment-1579469057
Since we support aarch64 in all Operating Systems we support x64 in, we need to keep the patch, but only for aarch64.

**Comment by andsel** [comment]
The 🔴 CI errors, once launched locally with
```
ci/docker_integration_tests.sh split 0
```

fails to start Logstash: 
```
      using logstash.yml and separate config file
        behaves like it can send 1000 documents to and index from the dlq

org.logstash.integration.RSpecTests > rspecTests STANDARD_ERROR
    /opt/logstash/qa/integration/framework/fixture.rb:59: warning: Passing safe_level with the 2nd argument of ERB.new is deprecated. Do not use it, and specify other arguments as keyword arguments.
    /opt/logstash/qa/integration/framework/fixture.rb:59: warning: Passing trim_mode with the 3rd argument of ERB.new is deprecated. Use keyword argument like ERB.new(str, trim_mode: ...) instead.

org.logstash.integration.RSpecTests > rspecTests STANDARD_OUT
    Starting Logstash: /opt/logstash/build/logstash-8.9.0-SNAPSHOT/bin/logstash ["-f", "/tmp/studtmp-153c90c79eb67c09d0029b9bfd8475923a7762f1960d4ba85e1932b9c78e", "--path.settings", "/tmp/studtmp-ac45a5f8ad2ee6d79594c605d7ec24c3b025edf830e4ad1d16fdbcfa2d80"] (pwd: /opt/logstash/qa/integration)
    Logstash started with PID 3408, using java: bundled java
    Using bundled JDK: /opt/logstash/build/logstash-8.9.0-SNAPSHOT/jdk
    [FATAL] 2023-06-20 18:25:45.455 [main] Logstash - Logstash was unable to start due to an unexpected Gemfile change.
    If you are a user, this is a bug.
    If you are a logstash developer, please try restarting logstash with the `--enable-local-plugin-development` flag set.
```

Also launching locally fails with same error, and happens that because the gradle build
```
./gradlew clean installDefaultGems
```
Fails with
```
ERROR: Installation Aborted, message: uninitialized constant Bundler::SolveFailure
```
that should be fixed by #15066

File: rubyUtils.gradle [hunk 4]
```diff
@@ -25,7 +25,7 @@ buildscript {
     dependencies {
         classpath "org.yaml:snakeyaml:${snakeYamlVersion}"
         classpath "de.undercouch:gradle-download-task:4.0.4"
-        classpath "org.jruby:jruby-core:9.3.10.0"
+        classpath "org.jruby:jruby-core:9.4.0.0"
```
Comment (jsvd) [inline rubyUtils.gradle:28]:
```suggestion
        classpath "org.jruby:jruby-core:9.4.3.0"
```

**Review by jsvd — APPROVED** [review]
LGTM
